chapters/functions.tex

```diff
@@ -2457,13 +2459,30 @@ \subsection{Utility Functions}\doublelabel{utility-functions}
 \lstinline!ModelicaError! or \lstinline!ModelicaFormatError! to signal the
 error.
 \end{tabular}\\ \hline
+\lstinline!ModelicaDuplicateString! &
+\begin{tabular}{@{}p{8.cm}@{}}
+\emph{char* ModelicaDuplicateString(const char*str)} \\
+Returns a writeable duplicate of the NUL-terminated string \lstinline!str!. 
+If an error occurs, this function does
+not return, but calls "\lstinline!ModelicaError!".
+\end{tabular}\\ \hline
+\lstinline!ModelicaDuplicateStringWithErrorReturn! &
+\begin{tabular}{@{}p{8.cm}@{}}
+\emph{char*\newline ModelicaDuplicateStringWithErrorReturn(const char*str)}\\
+Same as
+\lstinline!ModelicaDuplicateString!, except that in case of error, the function
+returns 0. This allows the external function to close files and free
+other open resources in case of error. After cleaning up resources use
+\lstinline!ModelicaError! or \lstinline!ModelicaFormatError! to signal the
+error.
+\end{tabular}\\ \hline
```

Comment:
I am missing this part:
> The duplicate shall be used as return argument of an external Modelica function.

The intention was to clarify how ownership is transferred out of the external function, as the external function itself has no way of releasing the memory.

Reply:
The formulation of `ModelicaAllocateString` is better in this respect, but it may not hurt to point out that the purpose of the rule is to take care of deallocation…

Reply:
I see, would it make sense to generalize it to:
The return values of these functions shall be used as return argument of an external Modelica function.

Reply:
…, thereby transferring ownership of the allocated memory to the caller of the external function.  _[There is no function for deallocating the memory, so returning it from the function is the only way to take care of its deallocation.]_

Reply:
You don't have to transfer ownership, since "12.9.1.1 Simple Types" states that you cannot store it for later use. Thus I would at least remove the non-normative part.
That is good since if you allocate two strings and the second allocation fails that shouldn't lead to problem.
However, this also implies that you could also use the string as a local buffer (e.g. when printing a message) - without returning it; and thus it might be best to not add this at all.

Reply:
OK, then I suppose this just shows that what is said in 12.9.1.1 really belongs to the section where the string allocation functions are defined.

What confuses me is that I the corresponding rule for `ModelicaAllocateString` was [clarified rather recently](https://github.com/modelica/ModelicaSpecification/issues/2279#issuecomment-442801105), and that I thought there were good reasons behind _removing "note"_.

Reply:
I agree moving that part from 12.9.1.1 simple types to the string allocation section would be good (especially now that there is a string allocation section). But I believe it should be a separate PR.

Reply:
However, I agree that it would make sense to move all of the allocation part from simple types to this new section. Should I do that to get the PR accepted?

Reply:
Yes, while some might call it scope creep, it would make sense to me to include it in this commit since the need for a better increases with the number of allocation functions.  I'll resolve this comment anyway, so you're free to do it here or as a separate PR.